cyto_dl/nn/vits/jepa.py

```diff
@@ -0,0 +1,192 @@
+        emb_dim: Optional[int] = 192,
+        num_layer: Optional[int] = 12,
+        num_head: Optional[int] = 3,
+        context_pixels: Optional[List[int]] = [0, 0, 0],
+        input_channels: Optional[int] = 1,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        num_patches: List[int]
+            Number of patches in each dimension
+        spatial_dims: int
+            Number of spatial dimensions
+        patch_size: List[int]
+            Size of each patch
+        emb_dim: int
+            Dimension of embedding
+        num_layer: int
+            Number of transformer layers
+        num_head: int
+            Number of heads in transformer
+        context_pixels: List[int]
+            Number of extra pixels around each patch to include in convolutional embedding to encoder dimension.
+        input_channels: int
+            Number of input channels
+        """
+        super().__init__()
+        if isinstance(num_patches, int):
+            num_patches = [num_patches] * spatial_dims
+        if isinstance(patch_size, int):
+            patch_size = [patch_size] * spatial_dims
+        self.patchify = Patchify(
+            patch_size, emb_dim, num_patches, spatial_dims, context_pixels, input_channels
+        )
+
+        self.transformer = torch.nn.Sequential(
+            *[Block(emb_dim, num_head) for _ in range(num_layer)]
+        )
+
+        self.layer_norm = torch.nn.LayerNorm(emb_dim)
+
+    def forward(self, img, patchify=True):
+        if patchify:
+            patches, _, _, _ = self.patchify(img, mask_ratio=0)
+        else:
+            patches = img
+        patches = rearrange(patches, "t b c -> b t c")
+        features = self.layer_norm(self.transformer(patches))
+        return features
+
+
+class JEPAPredictor(torch.nn.Module):
+    """Class for predicting target features from context embedding."""
+
+    def __init__(
+        self,
+        num_patches: List[int],
+        input_dim: Optional[int] = 192,
+        emb_dim: Optional[int] = 192,
+        num_layer: Optional[int] = 12,
+        num_head: Optional[int] = 3,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        num_patches: List[int]
+            Number of patches in each dimension
+        emb_dim: int
+            Dimension of embedding
+        num_layer: int
+            Number of transformer layers
+        num_head: int
+            Number of heads in transformer
+        """
+        super().__init__()
+        self.transformer = torch.nn.ParameterList(
+            [
+                CrossAttentionBlock(
+                    encoder_dim=emb_dim,
+                    decoder_dim=emb_dim,
+                    num_heads=num_head,
+                )
+                for _ in range(num_layer)
+            ]
+        )
+
+        self.mask_token = torch.nn.Parameter(torch.zeros(1, 1, emb_dim))
+        self.pos_embedding = torch.nn.Parameter(torch.zeros(np.prod(num_patches), 1, emb_dim))
```

Comment:
maybe use the general pos embedding function?

Reply:
updated!